### ISSUE TITLE
hline at y=0 appears after setting yscale to log

### ISSUE THREAD

**Issue by dstansby** [matplotlib/matplotlib#8783]
### Bug report

**Bug summary**

Calling `ax.axhline(0)` after setting the y-scale to log rescales the y-axis to make the horizontal line appear at 1e-300.

**Code for reproduction**

```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
ax.scatter([1], [1])
ax.set_yscale('log')
ax.axhline(0)
plt.show()
```

**Actual outcome**

![figure_1](https://user-images.githubusercontent.com/6197628/27333637-722e904a-55be-11e7-8104-5007c59920ed.png)
with the warning
```
/Users/dstansby/matplotlib/lib/matplotlib/axes/_base.py:3208: UserWarning: Attempted to set non-positive ylimits for log-scale axis; invalid limits will be ignored.
  'Attempted to set non-positive ylimits for log-scale axis; '
```

**Expected outcome**
I would expect the horizontal line to be ignored, and axis limits to remain as they were originally.

**Matplotlib version**

  * Operating System: OSX 10.12.5
  * Matplotlib Version: master branch installed using pip
  * Python Version: 3.6.1

**Comment by tacaswell** [comment]
On the other hand, you asked it to draw a line at 0 a it made a best-effort to do so.  What does a line at a negative position do?

**Comment by anntzer** [comment]
Do nothing, which is what happens when the call to scatter is not included.
OTOH this is fixed by #8537.

**Comment by dstansby** [comment]
Trying a line at a negative position (`ax.axhline(-1)`) results in exactly the same behavior as above.

Changing the code to
```python
import matplotlib.pyplot as plt

fig, ax = plt.subplots()
ax.scatter([1], [1])
plt.semilogy()
ax.axhline(0)
plt.show()
```
ignores the hline completely, and just shows the single scatter point as it was before. In my opinion I think this should be the behavior of `ax.set_yscale('log') too. At the very least they should be consistent which I think is what #8537 is for.

**Comment by dstansby** [comment]
Fixed by #8836
